ghsa/ghsa.go

```diff
@@ -0,0 +1,178 @@
+		AppFs:       afero.NewOsFs(),
+		Retry:       retry,
+		Client:      client,
+	}
+}
+
+func (c Config) Update() error {
+	log.Print("Fetching GithubSecurityAdvisory")
+
+	for _, ecosystem := range c.Ecosystems {
+		err := c.update(ecosystem)
+		if err != nil {
+			return xerrors.Errorf("failed to update github security advisory: %w", err)
+		}
+	}
+	return nil
+}
+
+func (c Config) update(ecosystem SecurityAdvisoryEcosystem) error {
+	log.Printf("Fetching GithubSecurityAdvisory: %s", ecosystem)
+
+	dir := filepath.Join(c.VulnListDir, ghsaDir, strings.ToLower(string(ecosystem)))
+	if err := os.RemoveAll(dir); err != nil {
+		return xerrors.Errorf("unable to remove github security advisory directory: %w", err)
+	}
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return xerrors.Errorf("failed to mkdir: %w", err)
+	}
+
+	ghsas, err := c.FetchGithubSecurityAdvisories(ecosystem)
+	if err != nil {
+		return xerrors.Errorf("failed to fetch github security advisory: %w", err)
+	}
+
+	ghsaJsonMap := make(map[string]GithubSecurityAdvisoryJson)
+	for _, ghsa := range ghsas {
+		ghsa.Package.Name = strings.TrimSpace(ghsa.Package.Name)
+
+		ghsaJson, ok := ghsaJsonMap[ghsa.Advisory.GhsaId+ghsa.Package.Name]
+		if ok {
+			va := VersionAdvisory{
+				FirstPatchedVersion:    ghsa.FirstPatchedVersion,
+				VulnerableVersionRange: ghsa.VulnerableVersionRange,
+			}
+			ghsaJson.VersionAdvisories = append(ghsaJson.VersionAdvisories, va)
+			ghsaJsonMap[ghsa.Advisory.GhsaId+ghsa.Package.Name] = ghsaJson
+
+		} else {
+			ghsaJsonMap[ghsa.Advisory.GhsaId+ghsa.Package.Name] = GithubSecurityAdvisoryJson{
+				Severity:  ghsa.Severity,
+				UpdatedAt: ghsa.UpdatedAt,
+				Package:   ghsa.Package,
+				Advisory:  ghsa.Advisory,
+				VersionAdvisories: []VersionAdvisory{
+					{
+						FirstPatchedVersion:    ghsa.FirstPatchedVersion,
+						VulnerableVersionRange: ghsa.VulnerableVersionRange,
+					},
+				},
+			}
+		}
+	}
+
+	bar := pb.StartNew(len(ghsaJsonMap))
+	for _, ghsaJson := range ghsaJsonMap {
+		dir := filepath.Join(c.VulnListDir, ghsaDir, strings.ToLower(string(ecosystem)), strings.Replace(ghsaJson.Package.Name, ":", "/", -1))
+		err := c.saveGSA(dir, ghsaJson.Advisory.GhsaId, ghsaJson)
+		if err != nil {
+			return xerrors.Errorf("failed to save github security advisory: %w", err)
+		}
+		bar.Increment()
+	}
+	bar.Finish()
+	return nil
+}
+
+func (c Config) FetchGithubSecurityAdvisories(ecosystem SecurityAdvisoryEcosystem) ([]GithubSecurityAdvisory, error) {
+	var getVulnerabilitiesQuery GetVulnerabilitiesQuery
+	var ghsas []GithubSecurityAdvisory
+	variables := map[string]interface{}{
+		"ecosystem": ecosystem,
+		"total":     graphql.Int(100),
+		"cursor":    (*githubql.String)(nil),
+	}
+	for {
+		var err error
+		for i := 0; i <= c.Retry; i++ {
+			if i > 0 {
+				wait := math.Pow(float64(i), 2) + float64(utils.RandInt()%10)
+				log.Printf("retry after %f seconds\n", wait)
+				time.Sleep(time.Duration(time.Duration(wait) * time.Second))
```

Comment:
This logic is currently untested, we could add a mock that fails on first try and passes subsequently.